File: csharp/Dson/DsonExtInt32.cs
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Dson;

public class DsonExtInt32 : DsonValue, IEquatable<DsonExtInt32>, IComparable<DsonExtInt32>, IComparable
{
    private readonly int _type;
    private readonly bool _hasVal; // 比较时放前面
    private readonly int _value;

    public DsonExtInt32(int type, int? value)
        : this(type, value ?? 0, value.HasValue) {
    }

    public DsonExtInt32(int type, int value, bool hasVal = true) {
        Dsons.CheckSubType(type);
        Dsons.CheckHasValue(value, hasVal);
        _type = type;
        _value = value;
        _hasVal = hasVal;
    }

    public override DsonType DsonType => DsonType.EXT_INT32;
    public int Type => _type;
    public bool HasValue => _hasVal;
    public int Value => _value;

    #region equals

    public bool Equals(DsonExtInt32? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _type == other._type && _value == other._value && _hasVal == other._hasVal;
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((DsonExtInt32)obj);
    }

    public override int GetHashCode() {
        return HashCode.Combine(_type, _value, _hasVal);
    }

    public static bool operator ==(DsonExtInt32? left, DsonExtInt32? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonExtInt32? left, DsonExtInt32? right) {
        return !Equals(left, right);
    }

    public int CompareTo(DsonExtInt32? other) {
        if (ReferenceEquals(this, other)) return 0;
        if (ReferenceEquals(null, other)) return 1;
        var typeComparison = _type.CompareTo(other._type);
        if (typeComparison != 0) return typeComparison;
        var hasValComparison = _hasVal.CompareTo(other._hasVal);
        if (hasValComparison != 0) return hasValComparison;
        return _value.CompareTo(other._value);
    }

    public int CompareTo(object? obj) {
        if (ReferenceEquals(null, obj)) return 1;
        if (ReferenceEquals(this, obj)) return 0;
        return obj is DsonExtInt32 other ? CompareTo(other) : throw new ArgumentException($"Object must be of type {nameof(DsonExtInt32)}");
    }

    public static bool operator <(DsonExtInt32? left, DsonExtInt32? right) {
        return Comparer<DsonExtInt32>.Default.Compare(left, right) < 0;
    }

    public static bool operator >(DsonExtInt32? left, DsonExtInt32? right) {
        return Comparer<DsonExtInt32>.Default.Compare(left, right) > 0;
    }

    public static bool operator <=(DsonExtInt32? left, DsonExtInt32? right) {
        return Comparer<DsonExtInt32>.Default.Compare(left, right) <= 0;
    }

    public static bool operator >=(DsonExtInt32? left, DsonExtInt32? right) {
        return Comparer<DsonExtInt32>.Default.Compare(left, right) >= 0;
    }

    #endregion

    public override string ToString() {
        return $"{nameof(Type)}: {Type}, {nameof(Value)}: {Value}, {nameof(HasValue)}: {HasValue}, {nameof(DsonType)}: {DsonType}";
    }
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonExtInt32.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.types.ExtInt32;

import javax.annotation.Nonnull;
import java.util.Objects;

/**
 * long值的简单扩展
 *
 * @author wjybxx
 * date - 2023/4/19
 */
public final class DsonExtInt32 extends DsonValue implements Comparable<DsonExtInt32> {

    private final ExtInt32 extInt32;

    public DsonExtInt32(int type, int value) {
        this(new ExtInt32(type, value));
    }

    public DsonExtInt32(int type, Integer value) {
        this(new ExtInt32(type, value));
    }

    public DsonExtInt32(int type, int value, boolean hasValue) {
        this(new ExtInt32(type, value, hasValue));
    }

    public DsonExtInt32(ExtInt32 extInt32) {
        Objects.requireNonNull(extInt32);
        this.extInt32 = extInt32;
    }

    public ExtInt32 extInt32() {
        return extInt32;
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.EXT_INT32;
    }

    public int getType() {
        return extInt32.getType();
    }

    public int getValue() {
        return extInt32.getValue();
    }

    public boolean hasValue() {
        return extInt32.hasValue();
    }

    // region equals

    @Override
    public int compareTo(DsonExtInt32 o) {
        return extInt32.compareTo(o.extInt32);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DsonExtInt32 that = (DsonExtInt32) o;

        return extInt32.equals(that.extInt32);
    }

    @Override
    public int hashCode() {
        return extInt32.hashCode();
    }

    // endregion

    @Override
    public String toString() {
        return "DsonExtInt32{" +
                "extInt32=" + extInt32 +
                '}';
    }
}
